docs/constrained-generators/manual.md

```diff
@@ -121,7 +209,156 @@ dot to the end of the Haskell name, we add an underscore `(_) to the end of the
+
+  1.  `singleton_ :: (Ord a, HasSpec a) => Term a -> Term (Set a)`
+  2.  `union_ :: (Ord a, HasSpec a) => Term (Set a) -> Term (Set a) -> Term (Set a)`
+  3.  `subset_ :: (Ord a, HasSpec a) => Term (Set a) -> Term (Set a) -> Term Bool`
+  4.  `member_ :: (Ord a, HasSpec a) => Term a -> Term (Set a) -> Term Bool`
+  5.  `disjoint_ :: (Ord a, HasSpec a) => Term (Set a) -> Term (Set a) -> Term Bool`
+  6.  `fromList_ :: (Ord a, HasSpec a) => Term [a] -> Term (Set a)`
+
+###  Function symbols for Map
+
+`(HasSpec k, HasSpec v) => HasSpec (Map k v)`
+
+The function symbols of `(Map k v)` are:
+
+1.  `dom_ :: Ord k => Term (Map k v) -> Term (Set k)`
+2.  `rng_ :: Ord k => Term (Map k v) -> Term [v]`
+3.  `lookup_ :: Ord k => Term k -> Term (Map k v) -> Term (Maybe v)`
+
+## Generating from and checking against specifications
+
+Once we have written a `Specification` what can we do with it? Specifications have two interpretations.
+
+1.  We can interpret it as a generator of values the meet all the constraints inside the specification.
+2.  We can interpret it as a function that checks if a given value meets all the constraints inside the specification.
+
+
+The first interpretation of the specification is the function `genFromSpec`
+
+```
+--| Generate a value from the spec in the QuickCheck monad 'Gen'
+genFromSpec:: (HasCallStack, HasSpec a) => Specification a -> QuickCheck.Gen a
+```
+
+This function is very useful when writing QuickCheck properties. With it we can write 
+`Gen` instances that are not completely random, but instead meet a set of constraints.
+
+Consider a system of 4 variables _w,x,y,z_ where we want to test the QuickCheck *implication*  property
+`(w < x && x < y && y < z) ==> property (w < z)`
+We might write a QuickCheck property like this
+
+```
+prop1 :: Gen Property
+prop1 = do
+   (w,x,y,z) <- arbitrary :: Gen (Int,Int,Int,Int)
+   pure $ (w < x && x < y && y < z) ==> property (w < z)
+```
+
+The problem with this is that the probability that the condition `(w < x && x < y && y < z)` is True, for random
+`w`, `x`, `y`, and `z`, is pretty low, so the property will pass vacuously most of the time, making a poor test.
+We can observe this by
+
+```
+ghci> quickCheck prop1
+*** Gave up! Passed only 29 tests; 1000 discarded tests.
+```
+
+A vacuous pass, becomes a QuickCheck `discard`, so we cannot find 100 successful passes.
+We can do a better job by constraining the condition using `genFromSpec`
+
+```
+spec1 :: Specification (Int,Int,Int,Int) 
+spec1 = constrained' $ \ w x y z -> [w <. x, x <. y, y <. z]
+
+prop2:: Gen Property
+prop2 = do
+   (w,x,y,z) <- genFromSpec spec1
+   pure $ (w < x && x < y && y < z) ==> property (w < z)
+```
+
+Now the test passes.
+
+```
+ghci> quickCheck prop2
++++ OK, passed 100 tests.
+```
+
+Now this isn't a very good test either, since the precedent is alway true. A better solution would be to
+generate a mix, where the precedent is True most of the time, but sometimes False.
```

Comment:
C.f. my other comment about this not being true.

Reply:
Yes, I'd like to make this better. Lets talk about that.

Reply:
perils of overconstraining
using classify to observe this
using cover.